helix-core/src/main/java/org/apache/helix/task/TaskStateModelFactory.java

```diff
@@ -44,39 +48,29 @@
   private final ScheduledExecutorService _taskExecutor;
   private final ScheduledExecutorService _timerTaskExecutor;
   private ThreadPoolExecutorMonitor _monitor;
-  public final static int TASK_THREADPOOL_SIZE = 40;
 
   public TaskStateModelFactory(HelixManager manager, Map<String, TaskFactory> taskFactoryRegistry) {
-    this(manager, taskFactoryRegistry,
-        Executors.newScheduledThreadPool(TASK_THREADPOOL_SIZE, new ThreadFactory() {
-          private AtomicInteger threadId = new AtomicInteger(0);
-
-          @Override
-          public Thread newThread(Runnable r) {
-            return new Thread(r, "TaskStateModelFactory-task_thread-" + threadId.getAndIncrement());
-          }
-        }));
+    _manager = manager;
+    _taskFactoryRegistry = taskFactoryRegistry;
+    // TODO: revisit the logic here; we are creating a connection although we already have a manager
+    ConfigAccessor configAccessor = createConfigAccessor();
+    int threadPoolSize = TaskUtil.getTargetThreadPoolSize(configAccessor, _manager.getClusterName(),
+        _manager.getInstanceName());
+    configAccessor.close();
+    _taskExecutor = createTaskExecutor(threadPoolSize);
+    _timerTaskExecutor = createTimerTaskExecutor();
+    initializeTaskMonitor();
   }
 
+  // FIXME: DO NOT USE! This size of provided thread pool will not be reflected to controller
+  // properly, the controller may over schedule tasks to this participant.
   public TaskStateModelFactory(HelixManager manager, Map<String, TaskFactory> taskFactoryRegistry,
```

Comment:
Deprecate?

And if anyone is using this method, what is the current thread pool size in the live instance node?

Reply:
@jiajunwang 
I shared this thought with Neal offline, but let me also share it here.

The reason I think it would be more appropriate to just add a FIXME or a warning here is because there's a slim chance that we might need to use this constructor. This might be useful later when we want to make the thread pool size change take place dynamically, without requiring a JVM restart.

I didn't think it'd be a good idea to deprecate and undeprecate public APIs in the future, so I've asked Neal to put a FIXME here. If you think otherwise, let us know.

Reply:
@jiajunwang To answer your question there will be no current thread pool size defined, so it will come out as the default value when accessed.

Reply:
That won't be the case. If we have a dynamic pool size later, the change will be done inside this class (configuration, or some new defined Helix executors) instead of being exposed as a public constructor, I guess.

Reply:
If it'll be us using it, how about we make it private? Users shouldn't gain access to it anyways, right?

Reply:
That could be another option. The point here is to prevent users from using it in order to retain control. But I'm not sure it's possible to make it private - isn't this required by the interface?

Reply:
Deprecating it as resolved offline.